File: windows/Model/GameModel.cs
namespace SequenceIQ.Model;

public enum Difficulty { Easy, Hard, Master }
public enum Screen { Splash, Home, Difficulty, Levels, Game, Complete, GameOver, Settings, Shop, Achievements, Stats, ComingSoon }
public sealed record Puzzle(Difficulty Difficulty, int Level, int?[] Sequence, int Answer, int[] Options);

public sealed class GameModel {
    public Screen Screen { get; set; } = Screen.Splash;
    public Difficulty Difficulty { get; set; } = Difficulty.Easy;
    public int Level { get; set; } = 1;
    public int Lives { get; set; } = 3;
    public int Hints { get; set; } = 3;
    public int HintStage { get; set; }
    public string Message { get; set; } = "";
    public HashSet<string> Completed { get; } = [];
    public int Attempts { get; private set; }
    public int CorrectAnswers { get; private set; }

    public bool Open(Difficulty d) => d == Difficulty.Easy || (d == Difficulty.Hard && CompletedCount(Difficulty.Easy) == 20) || (d == Difficulty.Master && CompletedCount(Difficulty.Hard) == 20);
    public bool CanPlay(Difficulty d, int l) => Open(d) && l is >= 1 and <= 20 && (l == 1 || Completed.Contains(Key(d, l - 1)));
    public int CompletedCount(Difficulty d) => Completed.Count(x => x.StartsWith($"{d}-", StringComparison.Ordinal));
    public bool IsCompleted(Difficulty d, int l) => Completed.Contains(Key(d,l));
    private static string Key(Difficulty d, int l) => $"{d}-{l}";

    public Puzzle GetPuzzle() => PuzzleRepository.Get(Difficulty, Level);
    public void Start(Difficulty d, int l) { if (!CanPlay(d,l)) return; Difficulty=d; Level=l; Lives=3; HintStage=0; Message=""; Screen=Screen.Game; }
    public void Answer(int value) { if (Screen != Screen.Game) return; Attempts++; if (value == GetPuzzle().Answer) { CorrectAnswers++; Completed.Add(Key(Difficulty,Level)); Screen=Screen.Complete; } else { Lives--; Message=Lives==0 ? "Game Over" : "Try again."; if (Lives==0) Screen=Screen.GameOver; } }
    public void UseHint() { if (Hints>0 && HintStage<3) { Hints--; HintStage++; } }
    public void Retry() => Start(Difficulty,Level);
    public void Next() { if (Level<20 && IsCompleted(Difficulty,Level)) Start(Difficulty,Level+1); }
}

File: android/Model/GameModel.kt
package com.hindtechgroup.sequenceiq.Model

import android.content.Context
import com.hindtechgroup.sequenceiq.Helper.PuzzleRepository

enum class Difficulty(val title: String) { EASY("Easy"), HARD("Hard"), MASTER("Master") }
enum class Screen { SPLASH, HOME, DIFFICULTY, LEVELS, GAME, COMPLETE, GAME_OVER, SETTINGS, SHOP, ACHIEVEMENTS, STATS, COMING_SOON }
data class Puzzle(val difficulty: Difficulty,val level:Int,val sequence:List<Int?>,val answer:Int,val options:List<Int>)

class GameModel(context: Context) {
 private val prefs=context.getSharedPreferences("sequenceIQ",Context.MODE_PRIVATE); private val puzzles=PuzzleRepository.load(context)
 var screen=Screen.SPLASH; var difficulty=Difficulty.EASY; var level=1; var lives=3; var hints=prefs.getInt("hints",3); var hintStage=0; var message=""; var attempts=prefs.getInt("attempts",0); var correctAnswers=prefs.getInt("correct",0)
 private val completed=prefs.getStringSet("completed",emptySet())!!.toMutableSet()
 fun open(d:Difficulty)=d==Difficulty.EASY||(d==Difficulty.HARD&&completedCount(Difficulty.EASY)==20)||(d==Difficulty.MASTER&&completedCount(Difficulty.HARD)==20)
 fun canPlay(d:Difficulty,l:Int)=open(d)&&l in 1..20&&(l==1||completed.contains("${d.title}-${l-1}"))
 fun completed(d:Difficulty,l:Int)=completed.contains("${d.title}-$l"); fun completedCount(d:Difficulty)=completed.count{it.startsWith("${d.title}-")}
 fun puzzle():Puzzle=puzzles.firstOrNull{it.difficulty==difficulty&&it.level==level} ?: error("Puzzle not found: $difficulty $level")
 fun start(d:Difficulty,l:Int){if(!canPlay(d,l))return;difficulty=d;level=l;lives=3;hintStage=0;message="";screen=Screen.GAME}
 fun answer(value:Int){if(screen!=Screen.GAME)return;attempts++;if(value==puzzle().answer){correctAnswers++;completed.add("${difficulty.title}-$level");persist();screen=Screen.COMPLETE}else{lives--;message=if(lives==0)"Game Over" else "Try again.";persist();if(lives==0)screen=Screen.GAME_OVER}}
 fun useHint(){if(hints>0&&hintStage<3){hints--;hintStage++;persist()}};fun retry(){start(difficulty,level)};fun next(){if(level<20&&completed(difficulty,level))start(difficulty,level+1)}
 private fun persist(){prefs.edit().putInt("hints",hints).putInt("attempts",attempts).putInt("correct",correctAnswers).putStringSet("completed",completed).apply()}
}
